File: coding-test-1/practice-problems/solutions-python/11-most-common-letter.py
# Write a method that takes in a string. Your method should return the
# most common letter in the array, and a count of how many times it
# appears.
#
# Difficulty: medium.
import unittest


def most_common_letter(string):
    letters = {}
    for letter in list(string):
        if letter in letters:
            letters[letter] += 1
        else:
            letters[letter] = 1

    common_count = 0
    common_letter = None

    for letter, letter_count in letters.items():
        if letter_count > common_count:
            common_count = letter_count
            common_letter = letter

    return [common_letter, common_count]

# These are tests to check that your code is working. After writing
# your solution, they should all run.


class MostCommonLetter(unittest.TestCase):

    def test_three_numbers(self):
        self.assertEqual(most_common_letter('abca'), ['a', 2])

    def test_four_numbers(self):
        self.assertEqual(most_common_letter('abbab'), ['b', 3])

if __name__ == '__main__':
    unittest.main()

File: temp/coding-test-1/practice-problems/solutions/11-most-common-letter.rb
# Write a method that takes in a string. Your method should return the
# most common letter in the array, and a count of how many times it
# appears.
#
# Difficulty: medium.

def most_common_letter(string)
  most_common_letter = nil
  most_common_letter_count = nil

  idx1 = 0
  while idx1 < string.length
    letter = string[idx1]
    count = 0

    idx2 = 0
    while idx2 < string.length
      if string[idx2] == letter
        count += 1
      end
      idx2 += 1
    end

    if (most_common_letter_count == nil) || (count > most_common_letter_count)
      most_common_letter = letter
      most_common_letter_count = count
    end

    idx1 += 1
  end

  return [most_common_letter, most_common_letter_count]
end

# These are tests to check that your code is working. After writing
# your solution, they should all print true.

puts(
  'most_common_letter("abca") == ["a", 2]: ' +
  (most_common_letter('abca') == ['a', 2]).to_s
)
puts(
  'most_common_letter("abbab") == ["b", 3]: ' +
  (most_common_letter('abbab') == ['b', 3]).to_s
)
